### PR TITLE
[WIP] Add tests for issue #5340

### PR DESCRIPTION
**Pull request by tomasnorre**

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Test case looks good. Could you fix it as well?

**Comment by tomasnorre** [comment]
I'll try to. Thanks for confirming, that helps me a lot.

**Comment by tomasnorre** [comment]
I haven't found the issue/solution yet. Perhaps you can give me some pointers base on additional information. 

The https://github.com/rectorphp/rector/blob/master/packages/node-collector/src/NodeCollector/ParsedNodeCollector.php#L167 is called and the Car class is seen as `$node instanceof New_` where the `Vehicle`-class is seen as `$node instanceof Class_`. 

The https://github.com/rectorphp/rector/blob/d7353baf2c289b16c763219f46421d774a190f9a/rules/dead-code/src/UnusedNodeResolver/UnusedClassResolver.php#L120 only iterates over the `$this->parsedNodeCollector->getNews() as $newNode` which explains why the class is seen as to be removed.

As said, I don't know the solution to this, so if you have some pointers, I will be happy. I'll see if I can figure out what the problem is :)

File: rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * @var bool
+     */
+    protected $engine;
+
+    public function hasEngine(): bool
+    {
+        return $this->engine;
+    }
+}
```
Comment (TomasVotruba) [inline rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc:34]: This file should not be in `/Fixture/`, but in `/Source/` directory, as it's not the tested class

By convention, the test fixture should only contain tested class

**Comment by TomasVotruba** [comment]
Could you submit your WIP? It would be easier to give better feedback on real code

File: rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc (hunk 1 above)
Comment (tomasnorre) [reply rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc:34]: If I move the Vehicle class to source, then the tests are green.

**Comment by tomasnorre** [comment]
> Could you submit your WIP? It would be easier to give better feedback on real code

The code hasn't changed since the commit. Still trying to figure out what's going on.

**Comment by tomasnorre** [comment]
I'm currently trying to figure out if a function `isParent()` analog to `isAbstract()` would be a solution and how it should look like. 
It's not part of the Class_ implementation. But perhaps it should be. 

The info is there. 
Screenshot from 2021-02-08 20:08:58

File: rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc (hunk 1 above)
Comment (TomasVotruba) [reply rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc:34]: Ah, so it's not about class that extends, but class that has children, right?
I got confused by the fixture name.

In that case the `Vehicle` should be moved first, and named `SkipClassThatHasUsedChildren`. Then reflection check on:

```
$classReflection = $scope->getClassReflection();

if ($classReflection->getAncestors()) {
    return [];
}
```

Comment (tomasnorre) [reply rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc:34]: Then reflection check on:

```php
$classReflection = $scope->getClassReflection();

if ($classReflection->getAncestors()) {
    return [];
}
```

Not sure what you mean here? Where should this be done/checked?

Comment (TomasVotruba) [reply rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc:34]: Probably here https://github.com/rectorphp/rector/blob/9f6834c46ea7599446dce808d21a069a4a6a9189/rules/dead-code/src/Rector/Class_/RemoveUnusedClassesRector.php#L118

**Comment by samsonasik** [comment]
@tomasnorre I cherry-picked your commit at PR https://github.com/rectorphp/rector/pull/5498

**Pull request closed without merging**